File: app/schemas/weather.py
from pydantic import BaseModel

from app.models.domain import (
    CurrentWeatherReport,
    DailyForecastReport,
    HourlyForecastReport,
)


class CurrentWeather(BaseModel):
    temperature_c: float
    apparent_temperature_c: float
    relative_humidity_percent: int
    precipitation_probability_percent: int | None = None
    precipitation_mm: float
    rain_mm: float
    showers_mm: float | None = None
    weather_code: int
    weather_description: str
    wind_speed_kmh: float
    wind_direction_degrees: int | None = None
    uv_index: float
    pressure_hpa: float | None = None
    visibility_meters: float | None = None
    cloud_cover_percent: int | None = None
    time: str
    is_day: bool | None = None


class CurrentWeatherResponse(BaseModel):
    city: str
    source: str = "city"
    location_name: str
    display_name: str
    short_display_name: str | None = None
    administrative_levels: dict[str, str | None] | None = None
    country: str
    latitude: float
    longitude: float
    timezone: str
    accuracy_meters: float | None = None
    location_confidence: str = "exact"
    location_provider: str = "fallback"
    provider: str = "open_meteo"
    fallback_provider_used: bool = False
    fallback_provider: str | None = None
    needs_user_confirmation: bool = False
    location_candidates: list[str] = []
    current: CurrentWeather

    @classmethod
    def from_domain(cls, report: CurrentWeatherReport) -> "CurrentWeatherResponse":
        location = report.location
        return cls(
            city=location.city,
            source=location.source,
            location_name=location.location_name or location.city,
            display_name=location.display_name or location.location_name or location.city,
            short_display_name=location.short_display_name,
            administrative_levels=location.administrative_levels,
            country=location.country,
            latitude=location.latitude,
            longitude=location.longitude,
            timezone=location.timezone,
            accuracy_meters=location.accuracy_meters,
            location_confidence=location.location_confidence,
            location_provider=location.location_provider,
            needs_user_confirmation=location.needs_user_confirmation,
            location_candidates=location.location_candidates or [],
            provider=report.provider,
            fallback_provider_used=report.fallback_provider_used,
            fallback_provider=report.fallback_provider,
            current=CurrentWeather(
                temperature_c=report.current.temperature_c,
                apparent_temperature_c=report.current.apparent_temperature_c,
                relative_humidity_percent=report.current.relative_humidity_percent,
                precipitation_probability_percent=report.current.precipitation_probability_percent,
                precipitation_mm=report.current.precipitation_mm,
                rain_mm=report.current.rain_mm,
                showers_mm=report.current.showers_mm,
                weather_code=report.current.weather_code,
                weather_description=report.current.weather_description,
                wind_speed_kmh=report.current.wind_speed_kmh,
                wind_direction_degrees=report.current.wind_direction_degrees,
                uv_index=report.current.uv_index,
                pressure_hpa=report.current.pressure_hpa,
                visibility_meters=report.current.visibility_meters,
                cloud_cover_percent=report.current.cloud_cover_percent,
                time=report.current.time,
                is_day=report.current.is_day,
            ),
        )


class HourlyWeather(BaseModel):
    time: str
    temperature_c: float
    apparent_temperature_c: float
    relative_humidity_percent: int
    precipitation_probability_percent: int
    precipitation_mm: float
    rain_mm: float
    weather_code: int
    weather_description: str
    wind_speed_kmh: float
    uv_index: float
    is_day: bool | None = None


class HourlyForecastResponse(BaseModel):
    city: str
    source: str = "city"
    location_name: str
    display_name: str
    short_display_name: str | None = None
    administrative_levels: dict[str, str | None] | None = None
    country: str
    latitude: float
    longitude: float
    timezone: str
    accuracy_meters: float | None = None
    location_confidence: str = "exact"
    location_provider: str = "fallback"
    provider: str = "open_meteo"
    fallback_provider_used: bool = False
    fallback_provider: str | None = None
    needs_user_confirmation: bool = False
    location_candidates: list[str] = []
    hourly: list[HourlyWeather]

    @classmethod
    def from_domain(cls, report: HourlyForecastReport) -> "HourlyForecastResponse":
        return cls(
            city=report.location.city,
            source=report.location.source,
            location_name=report.location.location_name or report.location.city,
            display_name=report.location.display_name or report.location.location_name or report.location.city,
            short_display_name=report.location.short_display_name,
            administrative_levels=report.location.administrative_levels,
            country=report.location.country,
            latitude=report.location.latitude,
            longitude=report.location.longitude,
            timezone=report.location.timezone,
            accuracy_meters=report.location.accuracy_meters,
            location_confidence=report.location.location_confidence,
            location_provider=report.location.location_provider,
            needs_user_confirmation=report.location.needs_user_confirmation,
            location_candidates=report.location.location_candidates or [],
            provider=report.provider,
            fallback_provider_used=report.fallback_provider_used,
            fallback_provider=report.fallback_provider,
            hourly=[
                HourlyWeather(
                    time=item.time,
                    temperature_c=item.temperature_c,
                    apparent_temperature_c=item.apparent_temperature_c,
                    relative_humidity_percent=item.relative_humidity_percent,
                    precipitation_probability_percent=item.precipitation_probability_percent or 0,
                    precipitation_mm=item.precipitation_mm,
                    rain_mm=item.rain_mm,
                    weather_code=item.weather_code,
                    weather_description=item.weather_description,
                    wind_speed_kmh=item.wind_speed_kmh,
                    uv_index=item.uv_index,
                    is_day=item.is_day,
                )
                for item in report.hourly
            ],
        )


class DailyWeather(BaseModel):
    date: str
    weather_code: int
    weather_description: str
    temperature_max_c: float
    temperature_min_c: float
    precipitation_probability_max_percent: int
    rain_sum_mm: float
    wind_speed_max_kmh: float
    uv_index_max: float
    sunrise: str
    sunset: str


class DailyForecastResponse(BaseModel):
    city: str
    source: str = "city"
    location_name: str
    display_name: str
    short_display_name: str | None = None
    administrative_levels: dict[str, str | None] | None = None
    country: str
    latitude: float
    longitude: float
    timezone: str
    accuracy_meters: float | None = None
    location_confidence: str = "exact"
    location_provider: str = "fallback"
    provider: str = "open_meteo"
    fallback_provider_used: bool = False
    fallback_provider: str | None = None
    needs_user_confirmation: bool = False
    location_candidates: list[str] = []
    daily: list[DailyWeather]

    @classmethod
    def from_domain(cls, report: DailyForecastReport) -> "DailyForecastResponse":
        return cls(
            city=report.location.city,
            source=report.location.source,
            location_name=report.location.location_name or report.location.city,
            display_name=report.location.display_name or report.location.location_name or report.location.city,
            short_display_name=report.location.short_display_name,
            administrative_levels=report.location.administrative_levels,
            country=report.location.country,
            latitude=report.location.latitude,
            longitude=report.location.longitude,
            timezone=report.location.timezone,
            accuracy_meters=report.location.accuracy_meters,
            location_confidence=report.location.location_confidence,
            location_provider=report.location.location_provider,
            needs_user_confirmation=report.location.needs_user_confirmation,
            location_candidates=report.location.location_candidates or [],
            provider=report.provider,
            fallback_provider_used=report.fallback_provider_used,
            fallback_provider=report.fallback_provider,
            daily=[
                DailyWeather(
                    date=item.date,
                    weather_code=item.weather_code,
                    weather_description=item.weather_description,
                    temperature_max_c=item.temperature_max_c,
                    temperature_min_c=item.temperature_min_c,
                    precipitation_probability_max_percent=item.precipitation_probability_max_percent,
                    rain_sum_mm=item.rain_sum_mm,
                    wind_speed_max_kmh=item.wind_speed_max_kmh,
                    uv_index_max=item.uv_index_max,
                    sunrise=item.sunrise,
                    sunset=item.sunset,
                )
                for item in report.daily
            ],
        )


class SearchLocationResponse(BaseModel):
    city: str
    country: str
    latitude: float
    longitude: float
    timezone: str
    display_name: str
    short_display_name: str | None = None
    administrative_levels: dict[str, str | None] | None = None
    location_confidence: str
    location_provider: str

File: src/types/weather.ts
export type StudyShift = "morning" | "afternoon" | "evening";
export type VehicleType = "motorbike" | "bus" | "walking" | "car" | "bicycle";
export type AppView = "home" | "forecast" | "study" | "schedule" | "settings" | "auth";
export type StudyDateMode = "today" | "tomorrow" | "custom";
export type LocationMode = "current" | "search" | "confirmed";

export interface AdministrativeLevels {
  hamlet?: string | null;
  ward_or_commune?: string | null;
  district?: string | null;
  province?: string | null;
  country?: string | null;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
  accuracy?: number;
}

export type LocationQuery =
  | { mode: "search"; city: string }
  | { mode: "current"; latitude: number; longitude: number; accuracy?: number }
  | { mode: "confirmed"; latitude: number; longitude: number; displayName: string; shortDisplayName?: string; administrativeLevels?: AdministrativeLevels };

export interface StudySchedule {
  study_date: string;
  start_time: string;
  end_time: string;
  vehicle_type: VehicleType;
}

type BaseStudentAdviceRequest = StudySchedule;

export interface CityBasedAdviceRequest extends BaseStudentAdviceRequest {
  city: string;
  latitude?: never;
  longitude?: never;
}

export interface CoordinateBasedAdviceRequest extends BaseStudentAdviceRequest {
  city?: never;
  latitude: number;
  longitude: number;
  accuracy_meters?: number;
}

export type StudentAdviceRequest = CityBasedAdviceRequest | CoordinateBasedAdviceRequest;

export interface WeatherMetric {
  max_temperature_c: number;
  max_apparent_temperature_c: number;
  max_precipitation_probability_percent: number;
  total_rain_mm: number;
  max_wind_speed_kmh: number;
  max_uv_index: number;
  average_humidity_percent: number;
}

export interface HourlyForecastItem {
  time: string;
  temperature_c?: number;
  apparent_temperature_c?: number;
  relative_humidity_percent?: number;
  precipitation_probability_percent?: number;
  precipitation_mm?: number;
  rain_mm?: number;
  weather_code?: number;
  weather_description?: string;
  wind_speed_kmh?: number;
  uv_index?: number;
  is_day?: boolean | null;
}

export interface DailyForecastItem {
  date: string;
  weather_code: number;
  weather_description: string;
  temperature_max_c: number;
  temperature_min_c: number;
  precipitation_probability_max_percent: number;
  rain_sum_mm: number;
  wind_speed_max_kmh: number;
  uv_index_max: number;
  sunrise: string;
  sunset: string;
}

export interface StudentAdviceResponse {
  city: string;
  source?: "city" | "coordinates" | string;
  location_name: string;
  display_name?: string;
  short_display_name?: string | null;
  administrative_levels?: AdministrativeLevels | null;
  country: string;
  latitude: number;
  longitude: number;
  timezone?: string | null;
  accuracy_meters?: number | null;
  location_confidence?: "exact" | "district" | "province" | "coordinates" | string;
  location_provider?: string;
  provider?: string;
  fallback_provider_used?: boolean;
  fallback_provider?: string | null;
  provider_condition?: string | null;
  effective_condition?: string | null;
  override_source?: string | null;
  override_expires_at?: string | null;
  override_report_id?: string | null;
  override_intensity?: string | null;
  needs_user_confirmation?: boolean;
  location_candidates?: string[];
  study_date: string;
  start_time: string;
  end_time: string;
  vehicle_type: VehicleType;
  score: number;
  level: string;
  summary: string;
  timeline: StudyTimeline;
  metrics: WeatherMetric;
  recommendations: string[];
  warnings: string[];
  hourly_forecast: HourlyForecastItem[];
  weather_code: number;
  weather_description: string;
  is_day?: boolean | null;
  time: string;
  wind_speed_kmh: number;
  precipitation_probability_percent: number;
  temperature_c: number;
}

export interface BeforeAfterClassTimeline {
  time: string;
  message: string;
  temperature_c: number;
  precipitation_probability_percent: number;
  weather_description: string;
}

export interface DuringClassTimeline {
  time_range: string;
  message: string;
  max_temperature_c: number;
  max_precipitation_probability_percent: number;
}

export interface StudyTimeline {
  before_class: BeforeAfterClassTimeline;
  during_class: DuringClassTimeline;
  after_class: BeforeAfterClassTimeline;
}

export interface CurrentWeatherItem {
  temperature_c?: number;
  apparent_temperature_c?: number;
  relative_humidity_percent?: number;
  precipitation_probability_percent?: number | null;
  precipitation_mm?: number;
  rain_mm?: number;
  showers_mm?: number | null;
  weather_code?: number;
  weather_description?: string;
  wind_speed_kmh?: number;
  wind_direction_degrees?: number | null;
  uv_index?: number;
  pressure_hpa?: number | null;
  visibility_meters?: number | null;
  cloud_cover_percent?: number | null;
  time?: string;
  is_day?: boolean | null;
}

export interface CurrentWeatherResponse {
  city: string;
  source?: "city" | "coordinates" | string;
  location_name?: string;
  display_name?: string;
  short_display_name?: string | null;
  administrative_levels?: AdministrativeLevels | null;
  country: string;
  latitude: number;
  longitude: number;
  timezone: string;
  accuracy_meters?: number | null;
  location_confidence?: "exact" | "district" | "province" | "coordinates" | string;
  location_provider?: string;
  provider?: string;
  fallback_provider_used?: boolean;
  fallback_provider?: string | null;
  provider_condition?: string | null;
  effective_condition?: string | null;
  override_source?: string | null;
  override_expires_at?: string | null;
  override_report_id?: string | null;
  override_intensity?: string | null;
  provider_weather_code?: number | null;
  provider_weather_description?: string | null;
  needs_user_confirmation?: boolean;
  location_candidates?: string[];
  current: CurrentWeatherItem;
}

export type LocalWeatherCondition = "rain" | "no_rain" | "storm";
export type LocalWeatherIntensity = "light" | "moderate" | "heavy";

export interface LocalWeatherReportPayload {
  location_name: string;
  latitude: number;
  longitude: number;
  reported_condition: LocalWeatherCondition;
  intensity?: LocalWeatherIntensity | null;
  expires_in_minutes?: number;
}

export interface LocalWeatherReportResponse {
  id: string;
  user_id: string;
  location_name: string;
  latitude: number;
  longitude: number;
  reported_condition: LocalWeatherCondition;
  intensity: LocalWeatherIntensity | null;
  source: "user_report" | string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  expires_at: string;
}

export interface HourlyWeatherResponse {
  city: string;
  source?: "city" | "coordinates" | string;
  location_name?: string;
  display_name?: string;
  short_display_name?: string | null;
  administrative_levels?: AdministrativeLevels | null;
  country: string;
  latitude: number;
  longitude: number;
  timezone: string;
  accuracy_meters?: number | null;
  location_confidence?: "exact" | "district" | "province" | "coordinates" | string;
  location_provider?: string;
  provider?: string;
  fallback_provider_used?: boolean;
  fallback_provider?: string | null;
  needs_user_confirmation?: boolean;
  location_candidates?: string[];
  hourly: HourlyForecastItem[];
}

export interface DailyWeatherResponse {
  city: string;
  source?: "city" | "coordinates" | string;
  location_name?: string;
  display_name?: string;
  short_display_name?: string | null;
  administrative_levels?: AdministrativeLevels | null;
  country: string;
  latitude: number;
  longitude: number;
  timezone: string;
  accuracy_meters?: number | null;
  location_confidence?: "exact" | "district" | "province" | "coordinates" | string;
  location_provider?: string;
  provider?: string;
  fallback_provider_used?: boolean;
  fallback_provider?: string | null;
  needs_user_confirmation?: boolean;
  location_candidates?: string[];
  daily: DailyForecastItem[];
}

export interface WeatherDisplayState {
  weather_code?: number;
  weather_description?: string;
  is_day?: boolean | null;
  time?: string;
  wind_speed_kmh?: number;
  precipitation_probability_percent?: number;
  temperature_c?: number;
}

export interface SearchLocationCandidate {
  city: string;
  country: string;
  latitude: number;
  longitude: number;
  timezone: string;
  display_name: string;
  short_display_name?: string | null;
  administrative_levels?: AdministrativeLevels | null;
  location_confidence: string;
  location_provider: string;
}

export interface User {
  id: string;
  email: string | null;
  username: string | null;
  full_name: string | null;
  avatar_url: string | null;
  auth_provider: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface UserLocationResponse {
  id: string;
  user_id: string;
  label: string;
  display_name: string;
  short_display_name: string | null;
  latitude: number;
  longitude: number;
  source: string;
  administrative_levels: AdministrativeLevels | null;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface StudyScheduleResponse {
  id: string;
  user_id: string;
  title: string;
  study_date: string | null;
  start_time: string;
  end_time: string;
  vehicle_type: VehicleType;
  location_id: string | null;
  repeat_type: string;
  repeat_days: string[] | null;
  note: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface UserSettingsResponse {
  id: string;
  user_id: string;
  temperature_unit: "celsius" | "fahrenheit";
  theme_mode: "auto" | "light" | "dark";
  auto_refresh_enabled: boolean;
  notification_enabled: boolean;
  default_vehicle_type: VehicleType;
  default_location_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface UserNotification {
  id: string;
  user_id: string;
  schedule_id: string | null;
  type: string;
  title: string;
  message: string;
  channel: string; // email | in_app | browser
  status: string; // pending | sent | failed | read
  error_message: string | null;
  scheduled_for: string | null;
  sent_at: string | null;
  read_at: string | null;
  occurrence_key?: string | null;
  risk_level?: string | null;
  content_hash?: string | null;
  created_at: string;
  updated_at: string;
}

export interface TestNotificationResponse {
  notification_id: string;
  channels_attempted: string[];
  channels_succeeded: string[];
  channels_failed: string[];
  message: string;
}
